proposals/0001-dependency-injection.md

```diff
@@ -0,0 +1,527 @@
+
+```kotlin
+fun interface DependencyProviderContext {
+    fun <T> set(key: DependencyKey, value: suspend AsyncDependencyResolutionContext.() -> T)
+}
+fun interface AsyncDependencyResolutionContext {
+    suspend fun <T: Any> get(key: DependencyKey): T
+}
+```
+
+### Naming
+[naming]: #naming
+
+For the *dependencies* scope:
+
+| Alternative | Reason for avoiding                  |
+|-------------|--------------------------------------|
+| di          | Not immediately recognizable acronym |
+| inject      | confusing for Koin users             |
+| providers   | too abstract                         |
+
+For the *provide* function:
+
+| Alternative | Reason for avoiding                                    |
+|-------------|--------------------------------------------------------|
+| inject      | confusing for Koin users                               |
+| supply      | this suggests that there is some scarcity of instances |
+| declare     | too abstract and non-operative                         |
+| single      | awkward wording out of context                         |
+
+
+## Resolution
+[resolution]: #resolution
+
+In practice, we can expect the declaring modules to be few, and the resolving modules to be many.
+
+Here is an example of how to resolve a dependency from the `Application` scope:
+
+```kotlin
+fun Application.configure() {
+    val repository by resolve<Repository<User>>()
+}
+```
+
+And the function declarations would look like:
+
+```kotlin
+inline fun <reified T> Application.resolve(key: String? = null): Lazy<T> =
+    attributes[AttributeKey<DependencyInjectionService>("DI")].resolutionContext().resolve(key)
+
+inline fun <reified T> DependencyResolutionContext.resolve(key: String? = null): Lazy<T> =
+    get(DependencyKey.of(typeInfo<T>(), key))
+```
+
+Which is using a context interface that looks like:
+
+```kotlin
+fun interface DependencyResolutionContext {
+    fun <T: Any> get(key: DependencyKey): Lazy<T>
+}
+```
+
+### Naming
+[naming-1]: #naming-1
+
+For the *resolve* function:
+
+| Alternative | Reason for avoiding                                 |
+|-------------|-----------------------------------------------------|
+| get         | overused; conflicting; too general                  |
+| inject      | doesn't seem to be the correct verb for the context |
+| locate      | too specific; not immediately obvious what it does  |
+
+
+# Technical Details
+[technical-details]: #technical-details
+
+In practice, we can expect developers to include several dependency declaration modules, followed by many more 
+resolutions in the feature modules.  To handle this, the implementation will need to build a tree of dependencies 
+from the set of declaration blocks, then perform the resolutions in a blocking scope at the site of the first 
+"application-scoped" `resolve()` call.  This introduces temporal coupling between the two phases of declaration and 
+resolution, though the declarations can happen in any order.
```

Comment:
My thoughts are that during declaration they'll suspend like this, but at a certain point we have to decide that declarations are done so that we know when to fail.